panic: trying to decrypt invalid data: ciphertext too small

<!--
NOTE: Not filling out the issue template needs a good reason, otherwise the
issue may be closed instantly! Please take the time to help us debugging the
problem by collecting information, even if it seems irrelevant to you. Thanks!
-->

## Output of `restic version`

restic 0.6.1
compiled with go1.8.3 on windows/amd64

## How did you start restic exactly? (Include the complete command line)

`d:\restic\restic.exe -r d:\restic\backup rebuild-index`

## What backend/server/service did you use?

None

## Expected behavior

Be able to rebuild the index.

## Actual behavior

```
C:\>d:\restic\restic.exe -r d:\restic\backup rebuild-index
enter password for repository:
counting files in repo
panic: trying to decrypt invalid data: ciphertext too small

goroutine 40 [running]:
restic/crypto.Decrypt(0xc0420e20f0, 0xb78100, 0x0, 0x0, 0xb78100, 0x0, 0x0, 0x0, 0xc0441b50e0, 0xc042027000)
        src/restic/crypto/crypto.go:280 +0x662
restic/pack.List(0xc0420e20f0, 0xb26fe0, 0xc044185e00, 0x400000, 0x40, 0x400000, 0x0, 0x0, 0xc0421ea140)
        src/restic/pack/pack.go:228 +0xbf
restic/repository.(*Repository).ListPack(0xc042038de0, 0x3c7228555392a300, 0x63c3ac493f7c55c, 0xf66e7ddb526bc869, 0xe6fd42ae9401139e, 0x2, 0x2, 0xc0422dffb0, 0x2, 0x2, ...)
        src/restic/repository/repository.go:473 +0x17c
restic/list.AllPacks.func1(0x8cca20, 0xc0422d30c0, 0x0, 0x0, 0x0, 0x0, 0xc042230060, 0x8e2440, 0xc0421ea140, 0x0, ...)
        src/restic/list/list.go:42 +0xe7
restic/worker.(*Pool).runWorker(0xc0421ebe00, 0x7)
        src/restic/worker/pool.go:86 +0x206
created by restic/worker.New
        src/restic/worker/pool.go:40 +0x181
```

## Steps to reproduce the behavior

Ran a backup operation (in windows) to an external USB plugged drive, part way through, the USB cable 'apparently' became unplugged. Upon reconnecting the drive, running a 'restic check' prompted to run rebuild-index. This didn't go well.

## Do you have any idea what may have caused this?

See steps to reproduce.

d:\restic\restic -r d:\restic\backup rebuild-index
debug log file debug.log
debug enabled
enter password for repository:
counting files in repo
panic: trying to decrypt invalid data: ciphertext too small

goroutine 51 [running]:
restic/crypto.Decrypt(0xc0420de320, 0xd071d8, 0x0, 0x0, 0xd071d8, 0x0, 0x0, 0x0, 0xc04202e800, 0x5d)
        src/restic/crypto/crypto.go:280 +0x662
restic/pack.List(0xc0420de320, 0xcaec60, 0xc0441ecf30, 0x400000, 0x4, 0xc0441f65c0, 0x40, 0x400000, 0x0)
        src/restic/pack/pack.go:228 +0xbf
restic/repository.(*Repository).ListPack(0xc0420d09c0, 0xcb3a20, 0xc04203e300, 0x3c7228555392a300, 0x63c3ac493f7c55c, 0xf66e7ddb526bc869, 0xe6fd42ae9401139e, 0xc042280420, 0x6, 0x6, ...)
        src/restic/repository/repository.go:474 +0x199
restic/list.AllPacks.func1(0xcb3a20, 0xc04203e300, 0x9cb280, 0xc042342900, 0x0, 0x0, 0x0, 0x0, 0x9e7080, 0xc04200c7c0, ...)
        src/restic/list/list.go:43 +0x10a
restic/worker.(*Pool).runWorker(0xc042110000, 0xcb3a20, 0xc04203e300, 0x9)
        src/restic/worker/pool.go:86 +0x253
created by restic/worker.New
        src/restic/worker/pool.go:40 +0x15d

Ergh, still no good;

>restic version
debug log file debug.txt
debug enabled
restic 0.6.1 (v0.6.1-32-g1aab123-dirty)
compiled with go1.8 on windows/amd64

>restic -r d:\restic\backup rebuild-index
debug log file debug.txt
debug enabled
enter password for repository:
counting files in repo
panic: trying to decrypt invalid data: ciphertext too small

goroutine 47 [running]:
restic/crypto.Decrypt(0xc0420e0370, 0xd07220, 0x0, 0x0, 0xd07220, 0x0, 0x0, 0x0, 0xc04216a800, 0x5d)
        src/restic/crypto/crypto.go:280 +0x662
restic/pack.List(0xc0420e0370, 0xcaec60, 0xc0442e4090, 0x400000, 0x4, 0xc04203fd40, 0x40, 0x400000, 0x0)
        src/restic/pack/pack.go:254 +0xbf
restic/repository.(*Repository).ListPack(0xc0420d69c0, 0xcb3a60, 0xc04203e8c0, 0x3c7228555392a300, 0x63c3ac493f7c55c, 0xf66e7ddb526bc869, 0xe6fd42ae9401139e, 0xc0442b0160, 0x6, 0x6, ...)
        src/restic/repository/repository.go:474 +0x199
restic/list.AllPacks.func1(0xcb3a60, 0xc04203e8c0, 0x9cb660, 0xc044273120, 0x0, 0x0, 0x0, 0x0, 0x9e7460, 0xc0420d9680, ...)
        src/restic/list/list.go:43 +0x10a
restic/worker.(*Pool).runWorker(0xc042110060, 0xcb3a60, 0xc04203e8c0, 0x9)
        src/restic/worker/pool.go:86 +0x253
created by restic/worker.New
        src/restic/worker/pool.go:40 +0x15d

[debug.txt](https://github.com/restic/restic/files/1062065/debug.txt)

--- a/src/restic/pack/pack.go
+++ b/src/restic/pack/pack.go
@@ -8,6 +8,7 @@ import (
 	"restic"
 	"sync"
 
+	"restic/debug"
 	"restic/errors"
 
 	"restic/crypto"
@@ -195,23 +196,37 @@ var minFileSize = entrySize + crypto.Extension
 // readHeader reads the header at the end of rd. size is the length of the
 // whole data accessible in rd.
 func readHeader(rd io.ReaderAt, size int64) ([]byte, error) {
+	debug.Log("size: %v", size)
 	if size == 0 {
-		err := InvalidFileError{
-			Message: "file is empty",
-		}
+		err := InvalidFileError{Message: "file is empty"}
 		return nil, errors.Wrap(err, "readHeader")
 	}
 
 	if size < int64(minFileSize) {
-		err := InvalidFileError{
-			Message: "file is too small",
-		}
+		err := InvalidFileError{Message: "file is too small"}
 		return nil, errors.Wrap(err, "readHeader")
 	}
 
 	hl, err := readHeaderLength(rd, size)
 	if err != nil {
 		return nil, err
+	}
+
+	debug.Log("header length: %v", size)
+
+	if hl == 0 {
+		err := InvalidFileError{Message: "header length is zero"}
+		return nil, errors.Wrap(err, "readHeader")
+	}
+
+	if hl < crypto.Extension {
+		err := InvalidFileError{Message: "header length is too small"}
+		return nil, errors.Wrap(err, "readHeader")
+	}
+
+	if (hl-crypto.Extension)%uint32(entrySize) != 0 {
+		err := InvalidFileError{Message: "header length is invalid"}
+		return nil, errors.Wrap(err, "readHeader")
 	}
 
 	if int64(hl) > size-int64(binary.Size(hl)) {

--- a/src/restic/pack/pack.go
+++ b/src/restic/pack/pack.go
@@ -189,9 +189,26 @@ func readHeaderLength(rd io.ReaderAt, size int64) (uint32, error) {
 
 const maxHeaderSize = 16 * 1024 * 1024
 
+// we require at least one entry in the header, and one blob for a pack file
+var minFileSize = entrySize + crypto.Extension
+
 // readHeader reads the header at the end of rd. size is the length of the
 // whole data accessible in rd.
 func readHeader(rd io.ReaderAt, size int64) ([]byte, error) {
+	if size == 0 {
+		err := InvalidFileError{
+			Message: "file is empty",
+		}
+		return nil, errors.Wrap(err, "readHeader")
+	}
+
+	if size < int64(minFileSize) {
+		err := InvalidFileError{
+			Message: "file is too small",
+		}
+		return nil, errors.Wrap(err, "readHeader")
+	}
+
 	hl, err := readHeaderLength(rd, size)
 	if err != nil {
 		return nil, err
@@ -216,6 +233,15 @@ func readHeader(rd io.ReaderAt, size int64) ([]byte, error) {
 	}
 
 	return buf, nil
+}
+
+// InvalidFileError is return when a file is found that is not a pack file.
+type InvalidFileError struct {
+	Message string
+}
+
+func (e InvalidFileError) Error() string {
+	return e.Message
 }
 
 // List returns the list of entries found in a pack file.

